cmd: add the ability to specify localities in allocsim

BramGruneir opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/12516)
<!-- Reviewable:end -->

BramGruneir committed:
cmd: add the ability to specify localities in allocsim

petermattis commented:
Review status: 0 of 3 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[pkg/cmd/allocsim/main.go, line 196 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12516#-KZWdCM968xQOAasxLtB:-KZWdCM968xQOAasxLtC:buv86b9) ([raw file](https://github.com/cockroachdb/cockroach/blob/313f78dea3e9aaaf35ed00a0b405525c6af59bbe/pkg/cmd/allocsim/main.go#L196)):*
> ```Go
> 			loc = fmt.Sprintf(":%s", localities[i-1])
> 		}
> 		node := fmt.Sprintf("%d%s", i, loc)
> ```

Seems more natural to append to `node` if a locality exists:

```
node := fmt.Sprintf("%d")
if i <= len(localities) {
  node += fmt.Sprintf(":%s", localities[i-1])
}
```

---

*[pkg/cmd/allocsim/main.go, line 296 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12516#-KZWbTatiFwt5iQSkdWc:-KZWbTatiFwt5iQSkdWd:bwspuya) ([raw file](https://github.com/cockroachdb/cockroach/blob/313f78dea3e9aaaf35ed00a0b405525c6af59bbe/pkg/cmd/allocsim/main.go#L296)):*
> ```Go
> 	a := newAllocSim(c)
> 
> 	var localities [][]string
> ```

Somewhat confusing to have `localities` and `a.localities` that are so similar in name but have different contents. I think `s/localities/perNodeArgs/g` would make this clearer. 

This might be better as `map[int][]string` as the lookup of the per-node args then becomes: `args := perNodeArgs[i]` which works even if the map is `nil`.

---

*[pkg/cmd/allocsim/main.go, line 299 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12516#-KZWaSJ6D-D1WfU0fmDI:-KZWaSJ6D-D1WfU0fmDJ:b-h2vnqr) ([raw file](https://github.com/cockroachdb/cockroach/blob/313f78dea3e9aaaf35ed00a0b405525c6af59bbe/pkg/cmd/allocsim/main.go#L299)):*
> ```Go
> 	if *numLocalities > 0 {
> 		a.localities = make([]roachpb.Locality, len(c.Nodes), len(c.Nodes))
> 		// Add some localities to the cluster. The
> ```

Broken sentence: `The ?`.

---

*[pkg/cmd/allocsim/main.go, line 328 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12516#-KZWaJ4XJmv8XtJcUxwa:-KZWaJ4XJmv8XtJcUxwb:b-r67owy) ([raw file](https://github.com/cockroachdb/cockroach/blob/313f78dea3e9aaaf35ed00a0b405525c6af59bbe/pkg/cmd/allocsim/main.go#L328)):*
> ```Go
> 	}()
> 
> 	c.Start("allocsim", *workers, []string{}, flag.Args(), localities)
> ```

The third argument can be `nil` instead of an empty `[]string`.

---

*[pkg/cmd/internal/localcluster/localcluster.go, line 117 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12516#-KZWbx_jPzLy7OsUqJ-2:-KZWbx_jPzLy7OsUqJ-3:bqi6itg) ([raw file](https://github.com/cockroachdb/cockroach/blob/313f78dea3e9aaaf35ed00a0b405525c6af59bbe/pkg/cmd/internal/localcluster/localcluster.go#L117)):*
> ```Go
> 
> 	for i := range c.Nodes {
> 		nodeArgs := allNodeArgs
> ```

It is subtle that this works (and you should definitely understand why it is subtle). You should be making a copy of `allNodeArgs`:

```
nodeArgs := append([]string(nil), allNodeArgs...)
```

---

*[pkg/cmd/internal/localcluster/localcluster.go, line 120 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12516#-KZWbjAtUWx-O3IaXcv2:-KZWbjAtUWx-O3IaXcv3:b-ypf9by) ([raw file](https://github.com/cockroachdb/cockroach/blob/313f78dea3e9aaaf35ed00a0b405525c6af59bbe/pkg/cmd/internal/localcluster/localcluster.go#L120)):*
> ```Go
> 		if perNodeArgs != nil {
> 			nodeArgs = append(nodeArgs, perNodeArgs[i]...)
> 		}
> ```

See other comment about changing the type of `perNodeArgs` to `map[int][]string`. The above would then become:

```
nodeArgs := append([]string(nil), allNodeArgs...)
nodeArgs = append(nodeArgs, perNodeArgs[i]...)
```

---

*[pkg/cmd/internal/localcluster/localcluster.go, line 393 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12516#-KZWcLUGvfnqwRvdwLZI:-KZWcLUGvfnqwRvdwLZJ:b-kjg90h) ([raw file](https://github.com/cockroachdb/cockroach/blob/313f78dea3e9aaaf35ed00a0b405525c6af59bbe/pkg/cmd/internal/localcluster/localcluster.go#L393)):*
> ```Go
> 	}
> 
> 	log.Infof(context.Background(), "Starting Node: %s\n", n.args)
> ```

Did you mean to leave this in? The command line is already printed in `Node.Start`. 

Also `s/\n//g`. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12516)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Review status: 0 of 3 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[pkg/cmd/allocsim/main.go, line 196 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12516#-KZWdCM968xQOAasxLtB:-KZXZpbQAVOdG-1hfbYM:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/313f78dea3e9aaaf35ed00a0b405525c6af59bbe/pkg/cmd/allocsim/main.go#L196)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Seems more natural to append to `node` if a locality exists:

```
node := fmt.Sprintf("%d")
if i <= len(localities) {
  node += fmt.Sprintf(":%s", localities[i-1])
}
```
</blockquote></details>

Done.

---

*[pkg/cmd/allocsim/main.go, line 296 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12516#-KZWbTatiFwt5iQSkdWc:-KZY-eokeYfRxBrueIW4:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/313f78dea3e9aaaf35ed00a0b405525c6af59bbe/pkg/cmd/allocsim/main.go#L296)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Somewhat confusing to have `localities` and `a.localities` that are so similar in name but have different contents. I think `s/localities/perNodeArgs/g` would make this clearer. 

This might be better as `map[int][]string` as the lookup of the per-node args then becomes: `args := perNodeArgs[i]` which works even if the map is `nil`.
</blockquote></details>

Done.

---

*[pkg/cmd/allocsim/main.go, line 299 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12516#-KZWaSJ6D-D1WfU0fmDI:-KZXzxEVR3x-jeqNJIPD:b-o41sof) ([raw file](https://github.com/cockroachdb/cockroach/blob/313f78dea3e9aaaf35ed00a0b405525c6af59bbe/pkg/cmd/allocsim/main.go#L299)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Broken sentence: `The ?`.
</blockquote></details>

Just removed the comment entirely.

---

*[pkg/cmd/allocsim/main.go, line 328 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12516#-KZWaJ4XJmv8XtJcUxwa:-KZY-ULGs98604l0v2D3:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/313f78dea3e9aaaf35ed00a0b405525c6af59bbe/pkg/cmd/allocsim/main.go#L328)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

The third argument can be `nil` instead of an empty `[]string`.
</blockquote></details>

Done.

---

*[pkg/cmd/internal/localcluster/localcluster.go, line 117 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12516#-KZWbx_jPzLy7OsUqJ-2:-KZXZrp3vAA60tavu6pM:b-sftrpf) ([raw file](https://github.com/cockroachdb/cockroach/blob/313f78dea3e9aaaf35ed00a0b405525c6af59bbe/pkg/cmd/internal/localcluster/localcluster.go#L117)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

It is subtle that this works (and you should definitely understand why it is subtle). You should be making a copy of `allNodeArgs`:

```
nodeArgs := append([]string(nil), allNodeArgs...)
```
</blockquote></details>

Indeed.  It was bugging me.

---

*[pkg/cmd/internal/localcluster/localcluster.go, line 120 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12516#-KZWbjAtUWx-O3IaXcv2:-KZXzJI8jQ-GGtZVHyTo:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/313f78dea3e9aaaf35ed00a0b405525c6af59bbe/pkg/cmd/internal/localcluster/localcluster.go#L120)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

See other comment about changing the type of `perNodeArgs` to `map[int][]string`. The above would then become:

```
nodeArgs := append([]string(nil), allNodeArgs...)
nodeArgs = append(nodeArgs, perNodeArgs[i]...)
```
</blockquote></details>

Done.

---

*[pkg/cmd/internal/localcluster/localcluster.go, line 393 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12516#-KZWcLUGvfnqwRvdwLZI:-KZXzU7mpeTABN9vroU0:bn49qep) ([raw file](https://github.com/cockroachdb/cockroach/blob/313f78dea3e9aaaf35ed00a0b405525c6af59bbe/pkg/cmd/internal/localcluster/localcluster.go#L393)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Did you mean to leave this in? The command line is already printed in `Node.Start`. 

Also `s/\n//g`. 
</blockquote></details>

nope, left in from testing.  Removed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12516)*
<!-- Sent from Reviewable.io -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 3 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12516#-:-KZYJWv6YTHcCt7nQIue:bnfp4nl)*
<!-- Sent from Reviewable.io -->
